### DISCUSSION TITLE
Reinforcement Learning

### DISCUSSION THREAD

**Discussion by harshraj22** [harshraj22/DeepLearning#16] [Show and tell]
Benchmarking My Learning path for Reinforcement Learning.

Things I have idea about:
- [x] States, Actions, Policy, Values
- [x] There is something called belmann equation which is used for policy iteration, value iteration to solve RL problems
- [x] Neural Networks (The global function approximators) are used as policy in deep reinforcement learning models.
- [x] There is something called policy gradient, using which we can train Deep Neural Networks whose loss function need not be differentiable (ie. gradient is calculated without derivative). Not sure how correct I know now xp

P.S. Maintaining a references list [here](https://github.com/harshraj22/DeepLearning/blob/main/writeups/RL_Reference.md)

**Comment by harshraj22** [comment]
### Nov 28th, 2021

- Skimmed through the basics of Value Iteration and Q learning using [this blog](https://www.kaggle.com/charel/learn-by-example-reinforcement-learning-with-gym#Basic-Q-learning-algorithm) from kaggle.
- Would try out some of this on my own on Google Colab 😃 

To Do:
- Go through [Deep RL Course Blogs](https://simoninithomas.github.io/deep-rl-course/)

**Reply by harshraj22** [reply]
- [Value Iteration on Taxi-v3](https://github.com/harshraj22/DeepLearning/blob/main/implementations/ValueIteration/Value_Iteration_Taxi_v3.ipynb)

**Comment by harshraj22** [comment]
### Nov 29th, 2021

- Skimmed through the [RL crash course](https://www.youtube.com/watch?v=Mut_u40Sqz4). He was using pre-trained kind of models. Not too insightful.
- Went through [this](https://johnwlambert.github.io/policy-gradients/) blog about policy gradient. Not understood completely, but seems a good resource.
- Spent time understanding policy gradients from [ Pieter Abbeel's video](https://www.youtube.com/watch?v=AKbX1Zvo7r8&list=PLwRJQ4m4UJjNymuBM9RdmB3Z9N5-0IlY0&index=3&t=239s). See [this video](https://www.youtube.com/watch?v=cQfOQcpYRzE&t=1372s) for intution about the trajectories he is talking about. (Dynamics Model = Frozen lake kind of env, where you don't always get the action you want to take)
- went through various pytorch implementations of Monte Carlo Policy Gradient (The REINFORCE algorithm). Most of the things seem clear now, except the fact that at some places, authors use `torch.gather`, whose reason is unclear. Would try to dig deeper into this tomorrow, along with implementing it on one of the gym environments tomorrow.

**Comment by harshraj22** [comment]
### Nov30th, 2021
- went through [Pieter Abbeel's video](https://www.youtube.com/watch?v=AKbX1Zvo7r8&list=PLwRJQ4m4UJjNymuBM9RdmB3Z9N5-0IlY0&index=4&t=239s) video again. This time, things were more clear.
- No updates on `torch.gather`.
- Would finish up the vanilla policy gradient implementation by today. Going through [this](https://github.com/PacktPublishing/Hands-on-Reinforcement-Learning-with-PyTorch/blob/master/Section%204/4.3%20Policy%20Gradients%20REINFORCE.ipynb) implementation currently. Sounds insightful...
- Would look up in detail about value estimation methods before jumping on A2C or Actor Critic methods.

**Reply by harshraj22** [reply]
Some clarity about `torch.gather`[[1](https://www.datahubbs.com/reinforce-with-pytorch/)] from [this](https://towardsdatascience.com/an-intuitive-explanation-of-policy-gradient-part-1-reinforce-aa4392cbfd3c) post. the `gather` method helps in selecting the probabilities corresponding to the already chosen actions, as the Neural network (policy) returns the probabilities corresponding to all actions.

![image](https://user-images.githubusercontent.com/46635452/144103931-1fc4a940-ee7f-426f-b2c2-801a1eaf919a.png)
![image](https://user-images.githubusercontent.com/46635452/144104169-5ee460a7-8542-4e46-9e67-84b00594e254.png)

**Reply by harshraj22** [reply]
To Do:
- [ ] Read out [this](https://towardsdatascience.com/an-intuitive-explanation-of-policy-gradient-part-1-reinforce-aa4392cbfd3c) promising looking blog about "intuitive explanation about policy gradient"

**Comment by harshraj22** [comment]
### Dec 1st, 2021

![](https://c.tenor.com/pMz-9ArZRQoAAAAC/no-work-today-happy.gif)

**Comment by harshraj22** [comment]
### Dec 2nd, 2021

- Went through TRPO and PPO video from [Foundations of Deep RL Series](https://www.youtube.com/watch?v=KjWF8VIMGiY&t=16s) and [Deep RL bootcamp](https://www.youtube.com/playlist?list=PLAdk-EyP1ND8MqJEJnSvaoUShrAWYe51U). Did not understand completely.
- Would go through some additional blogs & articles about the same.

**Reply by harshraj22** [reply]
- Went through [this series of article](https://medium.com/swlh/coding-ppo-from-scratch-with-pytorch-part-2-4-f9d8b8aa938a) about implementing PPO using pytorch. Things became more clear, hopefully after a couple of implementation code, it would have more clarity.

**Comment by harshraj22** [comment]
### Dec 3rd, 2021

- Not much progress.
- watched videos on how to write better python code: by [Raymond Hettinger](https://www.youtube.com/watch?v=wf-BqAjZb8M)
- Would try to finish up Actor Critic By today

**Comment by harshraj22** [comment]
### Dec 4th, 2021

- Went through [Phil's](https://www.youtube.com/watch?v=hlv79rcHws0) video, implementing PPO in pytorch. things seem clear now.
- Would go through a few more implementations, and shift to Actor-Critic, DDPG etc soon.

**Reply by harshraj22** [reply]
A few points to note:
(Comparison across [Phil's](https://www.youtube.com/watch?v=hlv79rcHws0) implementation, and [another popular](https://medium.com/analytics-vidhya/coding-ppo-from-scratch-with-pytorch-part-3-4-82081ea58146) implementation)

- The way Advantage function is calculated differs. Seems some difference in the critic's loss as well.
- Need to understand [distributions](https://pytorch.org/docs/stable/distributions.html) package from pytorch.

Another popular [PPO implementation](https://nbviewer.org/github/MrSyee/pg-is-all-you-need/blob/master/02.PPO.ipynb), [Another](https://github.com/ashleve/deep_reinforcement_learning/blob/master/PPO.py)

<hr>
To Do: 

- [ ] Dig deep into the various Advantage functions, understanding the differences between the various implementations of PPO
- [ ] Clearly understand the difference between Monte Carlo (MC) and Temporal difference (TD)
- [ ] Read about Importance Sampling.

**Reply by harshraj22** [reply]
A set of [best practices](https://github.com/EmbersArc/PPO/blob/master/best-practices-ppo.md) to follow while implementing PPO

**Comment by harshraj22** [comment]
### Dec 5th, 2021

- Went through the basics of Advantage Actor Critic (A2C) through [this freecodecamp](https://www.freecodecamp.org/news/an-intro-to-advantage-actor-critic-methods-lets-play-sonic-the-hedgehog-86d6240171d/) blog. Doesn't seem too complex. Would go through some existing implementations to solidify my understanding about the same.

**Reply by harshraj22** [reply]
- Went through parts of [Pieter Abbeel's](https://www.youtube.com/watch?v=QASqaj_HUZw) video. Would try to finish it off tomorrow.

**Comment by harshraj22** [comment]
### Dec 6th, 2021

- Not Much Progress
- Going through [Self critical Sequence Training Paper](https://arxiv.org/pdf/1612.00563.pdf), trying to understand the Reinforcement learning part that I skipped earlier. Hopefully, now I will have a better understanding of all this 🤞

**Comment by harshraj22** [comment]
### Dec 7th, 2021

- No updates today. Would be on a small break.

**Comment by harshraj22** [comment]
### Dec 8th, 2021

- Break

**Comment by harshraj22** [comment]
### Dec 9th, 2021

- A good read about [RL in computer vision](https://arxiv.org/pdf/2108.11510.pdf)

**Comment by harshraj22** [comment]
### Dec 10th, 2021
- On a Break till 16th,
- Good Reads: [intro to Bayesian Optimization](https://distill.pub/2020/bayesian-optimization/)

**Comment by harshraj22** [comment]
### Dec 11th, 2021
- On a break

**Comment by harshraj22** [comment]
### Dec 12th, 2021
- On a break
- [Good Reads on RL overview and its usage in Computer Vision](https://arxiv.org/pdf/1810.06339.pdf)

**Comment by harshraj22** [comment]
### Dec 13th, 2021
- On a break

**Comment by harshraj22** [comment]
### Dec 14th, 2021
- On a break

**Comment by harshraj22** [comment]
### Dec 15th, 2021
- On a break

**Comment by harshraj22** [comment]
### Dec 16th, 2021
- On a break

**Comment by harshraj22** [comment]
### Dec 17th, 2021

- Went through REINFORCE algorithm again.
- Going through the code for PPO. Would try to finally finish off this implementation by tonight.

**Comment by harshraj22** [comment]
### Empty

**Comment by harshraj22** [comment]
### Dec 27th, 2021

- [reddit post](https://old.reddit.com/r/reinforcementlearning/comments/7s8px9/deep_reinforcement_learning_practical_tips/) for implementing and debugging RL
- [Best practices for PPO](https://github.com/llSourcell/Unity_ML_Agents/blob/master/docs/best-practices-ppo.md)

**Comment by harshraj22** [comment]
### Discussions on successfully Training the [PPO](https://github.com/harshraj22/DeepLearning/tree/main/implementations/PolicyBasedMethods/PPO)

- [this reddit comment](https://www.reddit.com/r/reinforcementlearning/comments/9d98f2/comment/e5h8pvn/?utm_source=share&utm_medium=web2x&context=3) mentions about batch size, and memory buffer size.
- [some other comments](https://andyljones.com/posts/rl-debugging.html) on debugging RL

**Reply by harshraj22** [reply]
[this](https://www.reddit.com/r/reinforcementlearning/comments/qkc1b1/comment/hj1o8ci/?utm_source=share&utm_medium=web2x&context=3) was a good. bug fix.
